### PR TITLE
Stop using ReleasableLock in Translog

### PR DESCRIPTION
**Pull request by original-brownbear**
As discussed in #107555 there's some overhead to the use of `ReleasableLock` and it's probably not a good idea using it on the hot(ish) path in this class.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 1]
```diff
@@ -680,23 +719,28 @@ public Snapshot newSnapshot(long fromSeqNo, long toSeqNo) throws IOException {
...
-                return current.read(location);
-            } else {
-                // read backwards - it's likely we need to read on that is recent
-                for (int i = readers.size() - 1; i >= 0; i--) {
-                    TranslogReader translogReader = readers.get(i);
-                    if (translogReader.generation == location.generation) {
-                        return translogReader.read(location);
+        try {
+            readLock.lock();
+            try {
```
Comment (mhl-b) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java:724]: Why use double try? Looks like previous version worked with single one.

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/translog/Translog.java:724]: True it's more code now, but that's the price to pay here for removing some overhead :)

Comment (mhl-b) [reply server/src/main/java/org/elasticsearch/index/translog/Translog.java:724]: Right :)
try-with-resources closes before catching exception

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 2]
```diff
@@ -1700,11 +1766,12 @@ public void trimUnreferencedReaders() throws IOException {
         } catch (final Exception ex) {
             closeOnTragicEvent(ex);
             throw ex;
+        } finally {
+            writeLock.unlock();
```
Comment (mhl-b) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java:1770]: does this require double try? or only with read lock

Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/translog/Translog.java:1770]: Only for the read-lock, the locks themselves are properly reentrant and we only acquire the write lock in `close OnTragicEvent`.

**Review by mhl-b — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks @mhl-b !

**Pull request merged**